src/server/resolve.rs

```diff
@@ -23,79 +22,89 @@ pub type Callback = Box<dyn FnOnce(Result<String>) + Send>;
 /// A trait for resolving store addresses.
 pub trait StoreAddrResolver: Send + Clone {
     /// Resolves the address for the specified store id asynchronously.
-    fn resolve(&self, store_id: u64, cb: Callback) -> Result<()>;
+    fn resolve(&self, _store_id: u64) -> BoxFuture<'_, Result<String>>;
 }
 
-/// A task for resolving store addresses.
-pub struct Task {
-    store_id: u64,
-    cb: Callback,
+struct StoreAddr {
+    addr: String,
+    last_update: Instant,
 }
 
-impl Display for Task {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        write!(f, "resolve store {} address", self.store_id)
-    }
+/// A store address resolver which is backed by a `PDClient`.
+pub struct PdStoreAddrResolver<T, RR>
+where
+    T: PdClient + 'static,
+    RR: RaftStoreRouter<RocksEngine> + 'static,
+{
+    inner: Arc<ResolverInner<T, RR>>,
 }
 
-struct StoreAddr {
-    addr: String,
-    last_update: Instant,
+impl<T, RR> Clone for PdStoreAddrResolver<T, RR>
+where
+    T: PdClient + 'static,
+    RR: RaftStoreRouter<RocksEngine> + 'static,
+{
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+        }
+    }
 }
 
-/// A runner for resolving store addresses.
-struct Runner<T, RR>
+struct ResolverInner<T, RR>
 where
-    T: PdClient,
-    RR: RaftStoreRouter<RocksEngine>,
+    T: PdClient + 'static,
+    RR: RaftStoreRouter<RocksEngine> + 'static,
 {
     pd_client: Arc<T>,
-    store_addrs: HashMap<u64, StoreAddr>,
+    store_addrs: futures::lock::Mutex<HashMap<u64, StoreAddr>>,
+    router: futures::lock::Mutex<RR>,
```

Comment:
Why use a Mutex for `RaftStoreRouter `?

Reply:
I think you could clone it for every thread

Reply:
Maybe cloning a router for every simgle resolve call is heavy. Is that acceptable?

Reply:
All future state machines produced by `resolve` share the same router in this PR.